File: Data Structures & Algorithms/decode-ways/submission-1.py
class Solution:
    def numDecodings(self, s: str) -> int:
        cache = [-1] * len(s)
        def dfs(i):
            if i >= len(s):
                return 1
            if s[i] == '0':
                return 0
            if cache[i] != -1:
                return cache[i]
            res = dfs(i+1)
            if i + 1 < len(s):
                if (s[i] == '1') or (s[i] == '2' and s[i+1] < '7'):
                    res += dfs(i+2)
            cache[i] = res
            return cache[i]

        return dfs(0)

File: Data Structures & Algorithms/coin-change-ii/submission-1.java
class Solution {
    int[][] memo;
    public int change(int amount, int[] coins) {
        memo = new int[coins.length][amount + 1];
        for(int[] row : memo) {
            Arrays.fill(row, -1);
        }
        Arrays.sort(coins); 
        return dfs(0, amount, coins);
    }

    private int dfs(int i, int amount, int[] coins) {
        if(amount == 0) return 1;
        if(i == coins.length) return 0;
        if(memo[i][amount] != -1) return memo[i][amount];

        int res = 0;
        if(amount >= coins[i]) {
            res = dfs(i + 1, amount, coins);
            res += dfs(i, amount - coins[i], coins);
        }
        return memo[i][amount] = res;
    }
}
